### PR TITLE
Add finalizer to link resource

### PR DESCRIPTION
**Pull request by NickCellino**
### Description

This adds a `Finalizer` to the Link controller. For now, this just handles deleting the `hcp-config` directory.

### Testing & Reproduction steps

In addition to automated unit tests, I performed the following manual test.

1. Run consul like so:
```
go run . agent -dev --log-level trace --data-dir /tmp/consul-test
```
2. Verify that `hcp-config` directory is created
```
➜  consul git:(nickcellino/CC-7338/add-link-finalizer) ls /tmp/consul-test
checkpoint-signature hcp-config           node-id              serf
➜  consul git:(nickcellino/CC-7338/add-link-finalizer) ls /tmp/consul-test/hcp-config
hcp-management-token server-config.json   successful-bootstrap
➜  consul git:(nickcellino/CC-7338/add-link-finalizer)
```

3. Create a Link:
```shell-session
➜  consul git:(nickcellino/CC-7338/add-link-finalizer) http PUT localhost:8500/api/hcp/v2/link/global data[resourceId]=$RESOURCE_ID data[clientSecret]=$CLIENT_SECRET data[clientId]=$CLIENT_ID X-Consul-Token:$CONSUL_HTTP_TOKEN
HTTP/1.1 200 OK
Content-Length: 617
Content-Type: text/plain; charset=utf-8
Date: Tue, 23 Jan 2024 21:27:06 GMT

{
    "data": {
        "clientId": "...",
        "clientSecret": "...",
        "resourceId": "organization/d95434ef-1f14-4b17-b23e-1eb608bb9fda/project/e263159f-b9ea-468e-8dc5-ee3d23403fe2/hashicorp.consul.global-network-manager.cluster/my-cluster"
    },
    "generation": "01HMW3GFRA65WXEG75PEQ4734K",
    "id": {
        "name": "global",
        "tenancy": {
            "peerName": "local"
        },
        "type": {
            "group": "hcp",
            "groupVersion": "v2",
            "kind": "Link"
        },
        "uid": "01HMW3GFRA65WXEG75PDFC5YQP"
    },
    "version": "29"
}
```

Verify from logs that finalizer has been added ✅ :
```
2024-01-23T16:27:06.763-0500 [TRACE] agent.server.controller-runtime: added finalizer: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HMW3GFRA65WXEG75PDFC5YQP\" name:\"global\" type:{group:\"hcp\" group_version:\"v2\" kind:\"Link\"} tenancy:{peer_name:\"local\"}"
```
4. Delete the Link:
```shell-session
➜  consul git:(nickcellino/CC-7338/add-link-finalizer) http DELETE localhost:8500/api/hcp/v2/link/global
HTTP/1.1 204 No Content
Date: Tue, 23 Jan 2024 21:32:45 GMT
```

Verify from logs that cleanup has been called ✅ :
```
2024-01-23T16:32:45.578-0500 [TRACE] agent.server.controller-runtime: cleaning up link resource: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HMW3GFRA65WXEG75PDFC5YQP\" name:\"global\" type:{group:\"hcp\" group_version:\"v2\" kind:\"Link\"} tenancy:{peer_name:\"local\"}"
2024-01-23T16:32:45.578-0500 [DEBUG] agent.server.controller-runtime: deleting hcp-config dir: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HMW3GFRA65WXEG75PDFC5YQP\" name:\"global\" type:{group:\"hcp\" group_version:\"v2\" kind:\"Link\"} tenancy:{peer_name:\"local\"}" dir=/tmp/consul-test/hcp-config
2024-01-23T16:32:45.579-0500 [TRACE] agent.server.controller-runtime: removed finalizer: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HMW3GFRA65WXEG75PDFC5YQP\" name:\"global\" type:{group:\"hcp\" group_version:\"v2\" kind:\"Link\"} tenancy:{peer_name:\"local\"}"
2024-01-23T16:32:45.579-0500 [TRACE] agent.server.controller-runtime: finally deleted: controller=consul.io/hcp/link managed_type=hcp.v2.Link resource-id="uid:\"01HMW3GFRA65WXEG75PDFC5YQP\" name:\"global\" type:{group:\"hcp\" group_version:\"v2\" kind:\"Link\"} tenancy:{peer_name:\"local\"}"
```
5. Verify that `hcp-config` directory no longer exists ✅
```
➜  consul git:(nickcellino/CC-7338/add-link-finalizer) ls /tmp/consul-test
checkpoint-signature node-id              serf
```

6. Verify that Link no longer exists:
```
➜  consul git:(nickcellino/CC-7338/add-link-finalizer) http GET localhost:8500/api/hcp/v2/link/global
HTTP/1.1 404 Not Found
Content-Length: 52
Content-Type: text/plain; charset=utf-8
Date: Tue, 23 Jan 2024 21:34:14 GMT

rpc error: code = NotFound desc = resource not found
```

### Links
https://hashicorp.atlassian.net/browse/CC-7338

### PR Checklist

* [x] updated test coverage
* [ ] external facing docs updated
* [ ] appropriate backport labels added
* [ ] not a security concern

### CODE REVIEW

**Comment by loshz** [comment]
What is the reasoning behind removing the HCP config dir on exit? What if a user wants to debug an issue?

File: internal/hcp/internal/controllers/link/controller.go [hunk 1]
```diff
@@ -7,17 +7,17 @@ import (
...
 
 	"github.com/hashicorp/consul/agent/hcp/bootstrap"
 	hcpclient "github.com/hashicorp/consul/agent/hcp/client"
 	"github.com/hashicorp/consul/agent/hcp/config"
 	"github.com/hashicorp/consul/internal/controller"
 	"github.com/hashicorp/consul/internal/hcp/internal/types"
+	"github.com/hashicorp/consul/internal/resource"
 	"github.com/hashicorp/consul/internal/storage"
 	pbhcp "github.com/hashicorp/consul/proto-public/pbhcp/v2"
 	"github.com/hashicorp/consul/proto-public/pbresource"
```
Comment (analogue) [inline internal/hcp/internal/controllers/link/controller.go:23]: FYI, `make tools` will install a pre-commit hook that automatically formats imports consistently on `git commit` 😄.

File: internal/hcp/internal/controllers/link/controller.go [hunk 2]
```diff
@@ -108,6 +108,21 @@ func (r *linkReconciler) Reconcile(ctx context.Context, rt controller.Runtime, r
 		return err
 	}
 
+	err = addFinalizer(ctx, rt, res)
+	if err != nil {
+		rt.Logger.Error("error adding finalizer to link resource", "error", err)
+		return err
+	}
```
Comment (analogue) [inline internal/hcp/internal/controllers/link/controller.go:115]: style nit:
```suggestion
	if err = addFinalizer(ctx, rt, res); err != nil {
		rt.Logger.Error("error adding finalizer to link resource", "error", err)
		return err
	}
```

File: internal/hcp/internal/controllers/link/controller.go [hunk 3]
```diff
@@ -108,6 +108,21 @@ func (r *linkReconciler) Reconcile(ctx context.Context, rt controller.Runtime, r
...
+
+	if resource.IsMarkedForDeletion(res) {
+		err := cleanup(ctx, rt, res, r.dataDir)
+		if err != nil {
+			rt.Logger.Error("error cleaning up link resource", "error", err)
+			return err
+		}
```
Comment (analogue) [inline internal/hcp/internal/controllers/link/controller.go:122]: same style nit as above

**Review by analogue — COMMENTED** [review]
Just a few minor things, but LGTM.

**Comment by mkam** [comment]
> What is the reasoning behind removing the HCP config dir on exit? What if a user wants to debug an issue?

The main issue is that when we load the configuration, we check if the hcp-config files exist first and load them locally instead of fetching from HCP. If we don't delete the directory and a user creates a new link resource with different values, then we would be loading the configuration associated with the deleted link instead of fetching the configuration for the new link. 

As a side note, we've considered changing our loading strategy to always attempt to fetch and then load locally as a fallback. So if that strategy is implemented, then we could potentially skip deleting the hcp-config.

The other more minor disadvantage is that if the user doesn't have any intention of reconnecting to HCP, then these files are just left behind.

**Review by mkam — APPROVED** [review]

**Comment by analogue** [comment]
You get a 🥇 for being the first team to use finalizers outside of core!  Nice work 😄

**Review by analogue — APPROVED** [review]

**Review by loshz — APPROVED** [review]

**Pull request merged**